### PR TITLE
[Flang][bbc] Prevent bbc -emit-fir command invoking all passes twice

### PR DESCRIPTION
**Pull request by agozillon**
Currently when the bbc tool is invoked with the emit-fir command the pass pipeline will be invoked twice causing passes added to the pass manager to run twice on the input IR.

This change seeks to prevent that from occuring by only invoking the pass managers run command once and situationally adding the HLFIR to FIR pass pipeline in the same scenario as before.

### CODE REVIEW

File: flang/tools/bbc/bbc.cpp [hunk 1]
```diff
@@ -389,18 +389,14 @@ static mlir::LogicalResult convertFortranSourceToMLIR(
...
+    // Add HLFIR to FIR lowering pass
+    if (emitFIR && useHLFIR)
       fir::createHLFIRToFIRPassPipeline(pm, llvm::OptimizationLevel::O2);
-      if (mlir::failed(pm.run(mlirModule))) {
-        llvm::errs() << "FATAL: lowering from HLFIR to FIR failed";
-        return mlir::failure();
-      }
+
+    if (mlir::failed(pm.run(mlirModule))) {
+      llvm::errs() << "FATAL: verification of FIR or HLFIR module failed";
```
Comment (tblah) [inline flang/tools/bbc/bbc.cpp:398]: Please keep the old error message as that is likely to be more descriptive about what is going on here

**Review by tblah — CHANGES_REQUESTED** [review]
Hi, thanks for looking into this.

Only the verifier pass is invoked twice. You can see this using `bbc --mlir-pass-statistics`.

I don't think this is the right fix. This patch would cause the verifier pass not to be run at all when `-emit-hlfir` is used. I think you should bring the second `pm.run(mlirModule)` out of the if statement.

**Comment by agozillon** [comment]
> --mlir-pass-statistics

Thank you for the option to check the passes being ran! It seems that it is restricted to the OpenMP passes being ran twice due to the location: https://github.com/llvm/llvm-project/blob/main/flang/tools/bbc/bbc.cpp#L378 so perhaps shifting the location these are added is the correct fix.

**Comment by agozillon** [comment]
Updated the PR, I opted to break off the OpenMP pass pipeline into it's own function/pass manager to be executed at the point where I believe it's required (after the module generation and prior to the first verification pass). 

I opted to do it this way as opposed to inserting a pass manager clear and then re-add of the verification pass to the pass manager in-between the runs as this hopefully makes the logic a little simpler to follow and makes any new additions easier as well (effectively keeping the OpenMP logic encased in it's own world people can for the most part ignore).

I am open to other alternatives though, but this seemed the simplest (that I could think of at least).

**Review by tblah — APPROVED** [review]
Thanks for the changes! Yes I think you have the OpenMP passes in the right place - they should run right after lowering.

I wonder if we should have a test for the OpenMP pass pipeline similar to `flang/test/Driver/mlir-pass-pipeline.f90`. I'm fine with that remaining out of the scope of this PR though as there apparently wasn't any testing before your changes.

**Comment by agozillon** [comment]
> Thanks for the changes! Yes I think you have the OpenMP passes in the right place - they should run right after lowering.
> 
> I wonder if we should have a test for the OpenMP pass pipeline similar to `flang/test/Driver/mlir-pass-pipeline.f90`. I'm fine with that remaining out of the scope of this PR though as there apparently wasn't any testing before your changes.

Thank you very much for your help with the PR and the review! Yes, we likely should have added a test like that and it's likely important to add if we end up with more OpenMP specific passes.

**Pull request merged**